internal/service/networkfirewall/firewall_data_source.go

```diff
@@ -0,0 +1,258 @@
+						},
+						"type": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+			"firewall_policy_arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"firewall_policy_change_protection": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+			"firewall_status": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"sync_states": {
+							Type:     schema.TypeSet,
+							Computed: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"availability_zone": {
+										Type:     schema.TypeString,
+										Computed: true,
+									},
+									"attachment": {
+										Type:     schema.TypeList,
+										Computed: true,
+										Elem: &schema.Resource{
+											Schema: map[string]*schema.Schema{
+												"endpoint_id": {
+													Type:     schema.TypeString,
+													Computed: true,
+												},
+												"subnet_id": {
+													Type:     schema.TypeString,
+													Computed: true,
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			"name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				AtLeastOneOf: []string{"arn", "name"},
+			},
+			"subnet_change_protection": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+			"subnet_mapping": {
+				Type:     schema.TypeSet,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"subnet_id": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+			"tags": tftags.TagsSchema(),
+			"update_token": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"vpc_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func dataSourceFirewallResourceRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*conns.AWSClient).NetworkFirewallConn
+	ignoreTagsConfig := meta.(*conns.AWSClient).IgnoreTagsConfig
+
+	input := &networkfirewall.DescribeFirewallInput{}
+
+	if v, ok := d.GetOk("arn"); ok {
+		input.FirewallArn = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("name"); ok {
+		input.FirewallName = aws.String(v.(string))
+	}
+
+	if input.FirewallArn == nil && input.FirewallName == nil {
+
+		return diag.FromErr(fmt.Errorf("must specify either arn, name, or both"))
+	}
+
+	output, err := conn.DescribeFirewallWithContext(ctx, input)
+	if !d.IsNewResource() && tfawserr.ErrCodeEquals(err, networkfirewall.ErrCodeResourceNotFoundException) {
+		log.Printf("[WARN] NetworkFirewall Firewall (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error reading NetworkFirewall Firewall (%s): %w", d.Id(), err))
```

Comment:
You can use `diag.Errorf` instead of `diag.FromErr(fmt.Errorf))`: https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/diag#Errorf

```suggestion
		return diag.Errorf("reading NetworkFirewall Firewall (%s): %s", d.Id(), err))
```